### ISSUE TITLE
Hairpin horizontal positioning enhancement

### ISSUE THREAD

**Issue by craigsapp** [rism-ch/verovio#1632]
The horizontal placement of hairpins can be improved.  Here is an example for illustration:

Original:

<img width="1495" alt="Screen Shot 2020-08-23 at 6 21 08 AM" src="https://user-images.githubusercontent.com/3487289/90979775-33d22600-e50c-11ea-8484-6fa5b9fa173d.png">

Verovio rendering:

<img width="1300" alt="Screen Shot 2020-08-23 at 6 17 27 AM" src="https://user-images.githubusercontent.com/3487289/90979773-30d73580-e50c-11ea-9420-b89d3434d323.png">

Notice that the endpoints of the hairpins do not line up at the same point in each example.  The verovio rendering always ends at the left side of a notehead, but this is visually confusing when the stem is on the right side of the note, so the hairpin appears to stop before the timestamp to which they are attached.  For example, the last hairpin visually appears to end between the last two eighth notes of the measure, but in actuality it is attached to the last eighth note and the stem placement on the other side of the note is what is causing the perception that it is placed between the notes rather than on the last note.

I propose the following rules for placement of the hairpin endpints:

(1) starting point is always at the left side of the notes (if there are any notes at the starting timestamp).

(2) for the ending timestamp, place the end of the hairpin to the furthest right stem of any note/chord at the ending timestamp.

For Grand Staff hairpins, the staff membership would need to be specified.  Is this implemented yet in verovio and/or MEI?   Here is the diagram of necessary `@places` for the Grand Staff:

<img width="329" alt="Screen Shot 2020-08-23 at 6 43 25 AM" src="https://user-images.githubusercontent.com/3487289/90980007-8d872000-e50d-11ea-8024-f620f8285b99.png">

For example, if the `<hairpin>` is at `@staff="1 2"` (position "C" in red), then all of the notes on all layers on the two staves would be checked for stems.   If the `<hairpin>` is placed at positions A, B, D, or E, then only the notes on that staff (or layer) would be checked.

A complication is that probably only the nearest layer(s) should be examined rather than all layers at the given timestamp.  This is because the horizontal alignment is not as sensitive to stems when they are far away, and closer stems have a higher proiority.

An ideal situation is that the starting/ending note-attachment points could be encoded as attributes on the hairpin, which I presume is not yet representable in MEI?  In other words, if an endpoint is attached at a timestamp, visual refinement of its horizontal position is needed as well, since this cannot be 100% determined algorithmically.  If there is no note-attachment attributes, then verovio would use the above algorithm; otherwise, it would attached to the left, center, or right of the notes at a given timestamp according to the refined horizontal offset.  There is `@ho`, but I would prefer a more symbolic description of the offset in terms of horizontal justification on the notehead, and then `@ho` could be used for further horizontal offset refinements beyond justification on noteheads.

MEI test data:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="4.0.0">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2020-08-23T06:28:12" version="3.0.0-dev-f93ad92">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="mdiv-0000002102575507">
    <score xml:id="score-0000000301674395">
     <scoreDef xml:id="scoredef-0000001520700087" midi.bpm="72">
      <staffGrp xml:id="staffgrp-0000000339679116" symbol="brace" bar.thru="true">
       <staffDef xml:id="staffdef-0000000046742069" n="1" lines="5">
        <clef xml:id="clef-L2F2" shape="G" line="2" />
        <keySig xml:id="keysig-L3F2" pname="g" mode="minor" sig="2f" />
        <meterSig xml:id="metersig-L5F2" count="9" unit="8" />
       </staffDef>
       <staffDef xml:id="staffdef-0000001189396055" n="2" lines="5">
        <clef xml:id="clef-L2F1" shape="F" line="4" />
        <keySig xml:id="keysig-L3F1" pname="g" mode="minor" sig="2f" />
        <meterSig xml:id="metersig-L5F1" count="9" unit="8" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1" n="25">
       <staff xml:id="staff-0000000598269673" n="1">
        <layer xml:id="layer-L1F2N1" n="1">
         <chord xml:id="chord-L8F2" dur="4">
          <note xml:id="note-L8F2S1" oct="4" pname="b" accid.ges="f" />
          <note xml:id="note-L8F2S2" oct="5" pname="d" accid.ges="n" />
         </chord>
         <chord xml:id="chord-L10F2" dur="8">
          <note xml:id="note-L10F2S1" oct="4" pname="d" accid.ges="n" />
          <note xml:id="note-L10F2S2" oct="4" pname="b" accid.ges="f" />
         </chord>
         <beam xml:id="beam-L11F2-L13F2">
          <chord xml:id="chord-L11F2" dur="8">
           <note xml:id="note-L11F2S1" oct="4" pname="e" accid.ges="f" />
           <note xml:id="note-L11F2S2" oct="5" pname="c" accid.ges="n" />
          </chord>
          <chord xml:id="chord-L12F2" dur="8">
           <note xml:id="note-L12F2S1" oct="4" pname="g" accid.ges="n" />
           <note xml:id="note-L12F2S2" oct="5" pname="e" accid.ges="f" />
          </chord>
          <chord xml:id="chord-L13F2" dur="8">
           <note xml:id="note-L13F2S1" oct="4" pname="c" accid.ges="n" />
           <note xml:id="note-L13F2S2" oct="4" pname="a" accid.ges="n" />
          </chord>
         </beam>
         <chord xml:id="chord-L14F2" dots="1" dur="4">
          <note xml:id="note-L14F2S1" oct="4" pname="d" accid.ges="n" />
          <note xml:id="note-L14F2S2" oct="4" pname="b" accid.ges="f" />
         </chord>
        </layer>
       </staff>
       <staff xml:id="staff-0000001876929047" n="2">
        <layer xml:id="layer-L1F1N1" n="1">
         <beam xml:id="beam-L8F1-L10F1">
          <note xml:id="note-L8F1" dur="8" oct="2" pname="f" accid.ges="n" />
          <note xml:id="note-L9F1" dur="8" oct="3" pname="f" accid.ges="n" />
          <note xml:id="note-L10F1" dur="8" oct="3" pname="f" accid.ges="n" />
         </beam>
         <beam xml:id="beam-L11F1-L13F1">
          <note xml:id="note-L11F1" dur="8" oct="2" pname="f" accid.ges="n" />
          <note xml:id="note-L12F1" dur="8" oct="3" pname="f" accid.ges="n" />
          <note xml:id="note-L13F1" dur="8" oct="3" pname="f" accid.ges="n" />
         </beam>
         <beam xml:id="beam-L14F1-L16F1">
          <note xml:id="note-L14F1" dur="8" oct="2" pname="b" stem.dir="up" accid.ges="f" />
          <note xml:id="note-L15F1" dur="8" oct="3" pname="f" stem.dir="up" accid.ges="n" />
          <note xml:id="note-L16F1" dur="8" oct="3" pname="f" stem.dir="up" accid.ges="n" />
         </beam>
        </layer>
       </staff>
       <hairpin xml:id="hairpin-L8F3" staff="1" tstamp="1.000000" tstamp2="0m+8.0000" form="dim" vgrp="100" />
       <slur xml:id="slur-L8F2-L13F2" staff="1" startid="#chord-L8F2" endid="#chord-L13F2" />
      </measure>
      <measure xml:id="measure-L17" n="26">
       <staff xml:id="staff-L17F2N1" n="1">
        <layer xml:id="layer-L17F2N1" n="1">
         <chord xml:id="chord-L18F2" dur="4">
          <note xml:id="note-L18F2S1" oct="4" pname="b" accid.ges="f" />
          <note xml:id="note-L18F2S2" oct="5" pname="d" accid.ges="n" />
         </chord>
         <chord xml:id="chord-L20F2" dur="8">
          <note xml:id="note-L20F2S1" oct="4" pname="d" accid.ges="n" />
          <note xml:id="note-L20F2S2" oct="4" pname="b" accid.ges="f" />
         </chord>
         <chord xml:id="chord-L21F2" dur="4" stem.dir="up">
          <note xml:id="note-L21F2S1" oct="4" pname="a" accid.ges="n" />
          <note xml:id="note-L21F2S2" oct="5" pname="c" accid.ges="n" />
         </chord>
         <chord xml:id="chord-L23F2" dur="8">
          <note xml:id="note-L23F2S1" oct="4" pname="b" accid="n" />
          <note xml:id="note-L23F2S2" oct="5" pname="d" accid.ges="n" />
         </chord>
         <chord xml:id="chord-L24F2" dur="4">
          <note xml:id="note-L24F2S1" oct="5" pname="c" accid.ges="n" />
          <note xml:id="note-L24F2S2" oct="5" pname="e" accid.ges="f" />
         </chord>
         <chord xml:id="chord-L26F2" dur="8">
          <note xml:id="note-L26F2S1" oct="5" pname="c" accid.ges="n" />
          <note xml:id="note-L26F2S2" oct="5" pname="e" accid.ges="f" />
         </chord>
        </layer>
       </staff>
       <staff xml:id="staff-L17F1N1" n="2">
        <layer xml:id="layer-L17F1N1" n="1">
         <beam xml:id="beam-L18F1-L20F1">
          <note xml:id="note-L18F1" dur="8" oct="2" pname="b" stem.dir="up" accid.ges="f" />
          <note xml:id="note-L19F1" dur="8" oct="3" pname="f" stem.dir="up" accid.ges="n" />
          <note xml:id="note-L20F1" dur="8" oct="3" pname="f" stem.dir="up" accid.ges="n" />
         </beam>
         <beam xml:id="beam-L21F1-L23F1">
          <note xml:id="note-L21F1" dur="8" oct="2" pname="f" accid.ges="n" />
          <note xml:id="note-L22F1" dur="8" oct="3" pname="f" accid.ges="n" />
          <note xml:id="note-L23F1" dur="8" oct="3" pname="f" accid.ges="n" />
         </beam>
         <beam xml:id="beam-L24F1-L26F1">
          <note xml:id="note-L24F1" dur="8" oct="2" pname="c" accid.ges="n" />
          <note xml:id="note-L25F1" dur="8" oct="3" pname="c" accid.ges="n" />
          <note xml:id="note-L26F1" dur="8" oct="3" pname="c" accid.ges="n" />
         </beam>
        </layer>
       </staff>
       <hairpin xml:id="hairpin-L18F3" staff="1" tstamp="1.000000" tstamp2="0m+9.0000" form="cres" vgrp="100" />
       <slur xml:id="slur-L18F2-L24F2" staff="1" startid="#chord-L18F2" endid="#chord-L24F2" />
      </measure>
      <measure xml:id="measure-L27" n="27">
       <staff xml:id="staff-L27F2N1" n="1">
        <layer xml:id="layer-L27F2N1" n="1">
         <chord xml:id="chord-L28F2" dur="4">
          <note xml:id="note-L28F2S1" oct="5" pname="c" accid.ges="n" />
          <note xml:id="note-L28F2S2" oct="5" pname="e" accid.ges="f" />
         </chord>
         <chord xml:id="chord-L30F2" dur="8">
          <note xml:id="note-L30F2S1" oct="4" pname="a" accid.ges="n" />
          <note xml:id="note-L30F2S2" oct="5" pname="c" accid.ges="n" />
         </chord>
         <beam xml:id="beam-L31F2-L33F2">
          <chord xml:id="chord-L31F2" dur="8">
           <note xml:id="note-L31F2S1" oct="4" pname="f" accid="s" />
           <note xml:id="note-L31F2S2" oct="4" pname="a" accid.ges="n" />
          </chord>
          <chord xml:id="chord-L32F2" dur="8">
           <note xml:id="note-L32F2S1" oct="4" pname="a" accid.ges="n" />
           <note xml:id="note-L32F2S2" oct="5" pname="c" accid.ges="n" />
          </chord>
          <chord xml:id="chord-L33F2" dur="8">
           <note xml:id="note-L33F2S1" oct="4" pname="f" accid.ges="s" />
           <note xml:id="note-L33F2S2" oct="4" pname="a" accid.ges="n" />
          </chord>
         </beam>
         <beam xml:id="beam-L34F2-L36F2">
          <chord xml:id="chord-L34F2" dur="8">
           <note xml:id="note-L34F2S1" oct="4" pname="g" accid.ges="n" />
           <note xml:id="note-L34F2S2" oct="4" pname="b" accid.ges="f" />
          </chord>
          <chord xml:id="chord-L35F2" dur="8">
           <note xml:id="note-L35F2S1" oct="4" pname="g" accid.ges="n" />
           <note xml:id="note-L35F2S2" oct="4" pname="b" accid.ges="f" />
          </chord>
          <chord xml:id="chord-L36F2" dur="8">
           <note xml:id="note-L36F2S1" oct="4" pname="a" accid.ges="n" />
           <note xml:id="note-L36F2S2" oct="5" pname="c" accid.ges="n" />
          </chord>
         </beam>
        </layer>
       </staff>
       <staff xml:id="staff-L27F1N1" n="2">
        <layer xml:id="layer-L27F1N1" n="1">
         <beam xml:id="beam-L28F1-L30F1">
          <note xml:id="note-L28F1" dur="8" oct="2" pname="c" accid.ges="n" />
          <note xml:id="note-L29F1" dur="8" oct="3" pname="c" accid.ges="n" />
          <note xml:id="note-L30F1" dur="8" oct="3" pname="c" accid.ges="n" />
         </beam>
         <beam xml:id="beam-L31F1-L33F1">
          <note xml:id="note-L31F1" dur="8" oct="2" pname="d" accid.ges="n" />
          <note xml:id="note-L32F1" dur="8" oct="3" pname="d" accid.ges="n" />
          <note xml:id="note-L33F1" dur="8" oct="2" pname="d" accid.ges="n" />
         </beam>
         <beam xml:id="beam-L34F1-L36F1">
          <note xml:id="note-L34F1" dur="8" oct="2" pname="g" accid.ges="n" />
          <chord xml:id="chord-L35F1" dur="8">
           <note xml:id="note-L35F1S1" oct="2" pname="g" accid.ges="n" />
           <note xml:id="note-L35F1S2" oct="3" pname="g" accid.ges="n" />
          </chord>
          <chord xml:id="chord-L36F1" dur="8">
           <note xml:id="note-L36F1S1" oct="2" pname="f">
            <accid xml:id="accid-L36F1S1" accid="n" func="caution" />
           </note>
           <note xml:id="note-L36F1S2" oct="3" pname="f">
            <accid xml:id="accid-L36F1S2" accid="n" func="caution" />
           </note>
          </chord>
         </beam>
        </layer>
       </staff>
       <hairpin xml:id="hairpin-L28F3" staff="1" tstamp="1.000000" tstamp2="0m+8.0000" form="dim" vgrp="100" />
       <slur xml:id="slur-L28F2-L33F2" staff="1" startid="#chord-L28F2" endid="#chord-L33F2" />
      </measure>
      <measure xml:id="measure-L37" n="28">
       <staff xml:id="staff-L37F2N1" n="1">
        <layer xml:id="layer-L37F2N1" n="1">
         <chord xml:id="chord-L38F2" dur="4">
          <note xml:id="note-L38F2S1" oct="4" pname="b" accid.ges="f" />
          <note xml:id="note-L38F2S2" oct="5" pname="d" accid.ges="n" />
         </chord>
         <chord xml:id="chord-L40F2" dur="8">
          <note xml:id="note-L40F2S1" oct="4" pname="d" accid.ges="n" />
          <note xml:id="note-L40F2S2" oct="4" pname="b" accid.ges="f" />
         </chord>
         <chord xml:id="chord-L41F2" dur="4">
          <note xml:id="note-L41F2S1" oct="4" pname="a" accid.ges="n" />
          <note xml:id="note-L41F2S2" oct="5" pname="c" accid.ges="n" />
         </chord>
         <chord xml:id="chord-L43F2" dur="8">
          <note xml:id="note-L43F2S1" oct="4" pname="b" accid="n" />
          <note xml:id="note-L43F2S2" oct="5" pname="d" accid.ges="n" />
         </chord>
         <chord xml:id="chord-L44F2" dur="4">
          <note xml:id="note-L44F2S1" oct="5" pname="c" accid.ges="n" />
          <note xml:id="note-L44F2S2" oct="5" pname="e" accid.ges="f" />
         </chord>
         <chord xml:id="chord-L46F2" dur="8">
          <note xml:id="note-L46F2S1" oct="5" pname="c" accid.ges="n" />
          <note xml:id="note-L46F2S2" oct="5" pname="e" accid.ges="f" />
         </chord>
        </layer>
       </staff>
       <staff xml:id="staff-L37F1N1" n="2">
        <layer xml:id="layer-L37F1N1" n="1">
         <beam xml:id="beam-L38F1-L40F1">
          <note xml:id="note-L38F1" dur="8" oct="2" pname="b" stem.dir="up" accid.ges="f" />
          <note xml:id="note-L39F1" dur="8" oct="3" pname="f" stem.dir="up" accid.ges="n" />
          <note xml:id="note-L40F1" dur="8" oct="3" pname="f" stem.dir="up" accid.ges="n" />
         </beam>
         <beam xml:id="beam-L41F1-L43F1">
          <note xml:id="note-L41F1" dur="8" oct="2" pname="f" accid.ges="n" />
          <note xml:id="note-L42F1" dur="8" oct="3" pname="f" accid.ges="n" />
          <note xml:id="note-L43F1" dur="8" oct="3" pname="f" accid.ges="n" />
         </beam>
         <beam xml:id="beam-L44F1-L46F1">
          <note xml:id="note-L44F1" dur="8" oct="2" pname="c" accid.ges="n" />
          <note xml:id="note-L45F1" dur="8" oct="3" pname="c" accid.ges="n" />
          <note xml:id="note-L46F1" dur="8" oct="3" pname="c" accid.ges="n" />
         </beam>
        </layer>
       </staff>
       <hairpin xml:id="hairpin-L38F3" staff="1" tstamp="1.000000" tstamp2="0m+9.0000" form="cres" vgrp="100" />
       <slur xml:id="slur-L38F2-L44F2" staff="1" startid="#chord-L38F2" endid="#chord-L44F2" />
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

**Comment by craigsapp** [comment]
Somewhat related to issue https://github.com/music-encoding/music-encoding/issues/705.

**Comment by craigsapp** [comment]
Position C is encoded as `@place="between"`/`@staff="1 2"` in MEI: https://github.com/music-encoding/guidelines/issues/182

**Comment by lpugin** [comment]
@craigsapp I implemented the `@place="between"`. With this value `@vgrp` is not necessary anymore. Actually, for now it is even ignore: See dynam-008 here https://www.verovio.org/test-suite.xhtml?cat=dynam if you add `@vgrp` elements placed in between will still be moved up (or down) if necessary, which I think would be the expected behavior. Otherwise, it would require space to be added between the staves, which would not look nice. 

Regarding the starting and ending points of hairpin, I am not sure what to do. If you use `@startid` en `@endid`, harpins will be aligned with the center of the note, which is not the same as the corresponding timestamp position. I think it would make sense to change those as you are suggesting. That is, when the start id points to a note, then make the hairpin start on the left point of the note, and when the endid is a note, make it end on the right point of the note. I think changing this with timestamp would cause problems.

**Comment by craigsapp** [comment]
> Regarding the starting and ending points of hairpin, I am not sure what to do. If you use `@startid` and `@endid`, harpins will be aligned with the center of the note, which is not the same as the corresponding timestamp position. I think it would make sense to change those as you are suggesting. That is, when the start id points to a note, then make the hairpin start on the left point of the note, and when the endid is a note, make it end on the right point of the note. I think changing this with timestamp would cause problems.

That is a good point.  If someone wants to refine the position of the hairpins (mostly the ending), then they could give the IDs of the note/chord to which it should be aligned.  If the stem of the note points in the direction of the hairpin, then the hairpin should go to the right side of the note, if the stem goes away then end at the left side of the hairpin.  If the endpoints are timestamps, then the endpoints would attached to the left sides of the notes by default (current behavior).

In other words, when the hairpin ending is attached to a specific note or chord, the haiprin end would stop at the same horizontal position as the stem.  This would also be more efficient, since you would not need to search for items at a given time and sort through them deciding on a placement for the hairpin ending.

At some point, a note-justfication parameter for hairpins (and probably text, dynamics as a general class as well for their starting points) would be useful.  There are some editorial style where the hairpin ends in the middle of a note rather than the left or right sides, and it is not always consistent in the same edition (but usually it is).

**Comment by craigsapp** [comment]
> `@craigsapp` I implemented the `@place="between"`. With this value `@vgrp` is not necessary anymore. Actually, for now it is even ignore: See dynam-008 here https://www.verovio.org/test-suite.xhtml?cat=dynam if you add `@vgrp` elements placed in between will still be moved up (or down) if necessary, which I think would be the expected behavior. Otherwise, it would require space to be added between the staves, which would not look nice.


That is great!  Did you implement it recently, or have I forgotten about it?  

A strange then happens when I try to create a PDF of that example in the MEI editor:

<img width="416" alt="Screen Shot 2020-08-28 at 2 38 05 PM" src="https://user-images.githubusercontent.com/3487289/91617173-4e593480-e93c-11ea-87bf-111094222942.png">

(Maybe a minimum fill height parameter similar to the one for justification of the system of a work might be useful).

The placement of the dynamics implies that the`@between` placement is assigned before the staves are vertically justified (ideally the dynamics would all be in the middle of this page :-).

**Comment by lpugin** [comment]
I just implemented it.

I suspect the spacing problem comes from the changes made the justification algorithm. I will double check it.

**Comment by lpugin** [comment]
Looks good in the editor now. Closing
![image](https://user-images.githubusercontent.com/689412/104956717-fd220e00-59cc-11eb-85f8-6796fcacde85.png)
